src/runtime/FunctionObject.cpp

```diff
@@ -270,54 +270,49 @@ NEVER_INLINE void FunctionObject::generateBytecodeBlock(ExecutionState& state)
     currentCodeSizeTotal += m_codeBlock->m_byteCodeBlock->memoryAllocatedSize();
 }
 
-Value FunctionObject::callSlowCase(ExecutionState& state, const Value& callee, const Value& receiver, const size_t argc, Value* argv, bool isNewExpression)
-{
-    if (LIKELY(callee.isObject())) {
-        if (LIKELY(callee.asPointerValue()->isFunctionObject()))
-            return callee.asFunction()->processCall(state, receiver, argc, argv, isNewExpression);
-#if ESCARGOT_ENABLE_PROXY_REFLECT
-        else if (callee.asPointerValue()->isProxyObject() && callee.asPointerValue()->isCallable())
-            return callee.asPointerValue()->asProxyObject()->call(state, callee, receiver, argc, argv);
-#endif
-    }
-    ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, errorMessage_Call_NotFunction);
-    return Value();
-}
-
-Object* FunctionObject::newInstance(ExecutionState& state, const size_t argc, Value* argv)
+// https://www.ecma-international.org/ecma-262/6.0/#sec-ecmascript-function-objects-construct-argumentslist-newtarget
+Object* FunctionObject::construct(ExecutionState& state, const size_t argc, Value* argv, const Value& newTarget)
 {
+    // FIXME: binded function
+    // FIXME: newTarget
     CodeBlock* cb = codeBlock();
-    FunctionObject* targetFunction = this;
+    FunctionObject* constructor = this;
 
-    if (UNLIKELY(cb->isBindedFunction())) {
+    if (UNLIKELY(cb->isBindedFunction() == true)) {
         // for nested bind function
-        while (cb->isBindedFunction()) {
-            targetFunction = Value(cb->boundFunctionInfo()->m_boundTargetFunction).asFunction();
-            cb = targetFunction->codeBlock();
+        while (cb->isBindedFunction() == true) {
+            constructor = Value(cb->boundFunctionInfo()->m_boundTargetFunction).asFunction();
+            cb = constructor->codeBlock();
         }
     }
 
-    if (UNLIKELY(!cb->isConstructor())) {
-        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, codeBlock()->functionName().string(), false, String::emptyString, errorMessage_New_NotConstructor);
+    // Assert: Type(newTarget) is Object.
+    ASSERT(newTarget.isObject() == true);
+    ASSERT(newTarget.isConstructor() == true);
+    // Let kind be F’s [[ConstructorKind]] internal slot.
+    ConstructorKind kind = constructorKind();
+    Object* thisArgument;
+
+    if (cb->hasCallNativeFunctionCode() == true) {
+        thisArgument = cb->nativeFunctionData()->m_ctorFn(state, cb, argc, argv);
+        // FIXME: If kind is "base", then
+    } else {
+        // FIXME: Let thisArgument be OrdinaryCreateFromConstructor(newTarget, "%ObjectPrototype%").
+        thisArgument = new Object(state);
     }
 
-    Object* receiver;
-    if (cb->hasCallNativeFunctionCode()) {
-        receiver = cb->nativeFunctionData()->m_ctorFn(state, cb, argc, argv);
+    if (constructor->getFunctionPrototype(state).isObject()) {
```

Comment:
For readability / State expression explicitly
if (constructor->getFunctionPrototype(state).isObject() == true)

Reply:
Fixed it. Thanks.